[AArch64] Add CodeGen support for scalar FEAT_CPA

rgwott opened the pull request:
CPA stands for Checked Pointer Arithmetic and is part of the 2023 MTE architecture extensions for A-profile.
The new CPA instructions perform regular pointer arithmetic (such as base register + offset) but check for overflow in the most significant bits of the result, enhancing security by detecting address tampering.

In this patch we intend to capture the semantics of pointer arithmetic when it is not folded into loads/stores, then generate the appropriate scalar CPA instructions. In order to preserve pointer arithmetic semantics through the backend, we use the PTRADD SelectionDAG node type.

Use backend option `-aarch64-use-featcpa-codegen=true` to enable CPA CodeGen (for a target with CPA enabled).

The story of this PR is that initially it introduced the PTRADD SelectionDAG node and the respective visitPTRADD() function, adapted from the CHERI/Morello LLVM tree. The original authors are @davidchisnall, @jrtc27, @arichardson.
After a while, @ritter-x2a took the part of the code that was target-independent and merged it separately in #140017. This PR thus remains as the AArch64-part only.

Mode details about the CPA extension can be found at:

- https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/arm-a-profile-architecture-developments-2023
- https://developer.arm.com/documentation/ddi0602/2023-09/ (e.g ADDPT instruction)

This PR follows #79569.
It does not address vector FEAT_CPA instructions.

rgwott committed:
[AArch64] Add CodeGen support for FEAT_CPA

CPA stands for Checked Pointer Arithmetic and is part of the 2023 MTE
architecture extensions for A-profile.
The new CPA instructions perform regular pointer arithmetic (such as
base register + offset) but check for overflow in the most significant
bits of the result, enhancing security by detecting address tampering.

In this patch we intend to capture the semantics of pointer arithmetic
when it is not folded into loads/stores, then generate the appropriate
CPA instructions. In order to preserve pointer arithmetic semantics
through the backend, we add the PTRADD SelectionDAG node type.

The PTRADD node and respective visitPTRADD() function are adapted from
the CHERI/Morello LLVM tree.

Mode details about the CPA extension can be found at:

- https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/arm-a-profile-architecture-developments-2023
- https://developer.arm.com/documentation/ddi0602/2023-09/

This PR follows #79569.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -4081,8 +4087,15 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
+        Addr = DAG.getNode(ISD::ADD, dl, PTy, Addr,
+                           TLI.getPICJumpTableRelocBase(Table, DAG));
+      } else {
+        // PTRADD always takes the pointer first, so the operands are commuted
```
You can just select on the opcode and emit them in the same order?

jrtc27 replied:
Yeah the only reason I did it this way downstream was to avoid test churn by commuting the ADD operands for non-CHERI targets

rgwott replied:
Emitting in the same order would break 20+ tests of several architectures. Although that is probably a simple problem, I do not believe fixing it should be in the scope of this PR. I changed `getMemBasePlusOffset()` to take an optional argument to allow the order of the operands to be inverted.

arichardson replied:
Tests changed in https://github.com/llvm/llvm-project/pull/125279

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -4081,8 +4087,15 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
       // RelocBase can be JumpTable, GOT or some sort of global base.
-      Addr = DAG.getNode(ISD::ADD, dl, PTy, Addr,
-                          TLI.getPICJumpTableRelocBase(Table, DAG));
+      if (!DAG.getTarget().shouldPreservePtrArith(
```
Probably should have this in a more convenient property in the DAG

arichardson replied:
Can't we use `getMemBasePlusOffset` here?

rgwott replied:
Done, using `getMemBasePlusOffset`.

arsenm commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -434,6 +434,11 @@ class TargetMachine {
+  /// True if target has some particular form of dealing with pointer arithmetic
+  /// semantics. False if pointer arithmetic should not be preserved for passes
+  /// such as instruction selection, and can fallback to regular arithmetic.
+  virtual bool shouldPreservePtrArith(const Function &F) const { return false; }
```
I'd expect this to be a property of individual instructions?

jrtc27 replied:
For us it's a property of the type, so having the instruction would suffice

jrtc27 replied:
Though getting the instruction in getMemBasePlusOffset isn't doable, and we really do need that interface to do the right thing whatever you throw at it. We add new `c<N>` types for capabilities that make it easy to know what you want to do, but CPA doesn't really fit in with that as they're still integers.

rgwott replied:
For FEAT_CPA this is a property of the target. Suggestions for improvement are welcome. =)

arsenm commented on llvm/lib/Target/AArch64/AArch64TargetMachine.cpp:
```diff
@@ -920,3 +920,7 @@ bool AArch64TargetMachine::parseMachineFunctionInfo(
+
+bool AArch64TargetMachine::shouldPreservePtrArith(const Function &F) const {
+  return getSubtargetImpl(F)->hasCPA();
+}
```
End of file error

rgwott replied:
Done.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -4069,8 +4069,14 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
+    SDValue Addr;
+    if (!DAG.getTarget().shouldPreservePtrArith(
+            DAG.getMachineFunction().getFunction())) {
+      Addr = DAG.getNode(ISD::ADD, dl, Index.getValueType(), Index, Table);
```
Keep the order the same and use `getMemBasePlusOffset`?

rgwott replied:
Done.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -4324,7 +4330,7 @@ void SelectionDAGBuilder::visitGetElementPtr(const User &I) {
           Flags.setNoUnsignedWrap(true);
 
-        N = DAG.getNode(ISD::ADD, dl, N.getValueType(), N,
+        N = DAG.getNode(AddOpcode, dl, N.getValueType(), N,
```
getMemBasePlusOffset?

rgwott replied:
Done.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -4368,7 +4374,7 @@ void SelectionDAGBuilder::visitGetElementPtr(const User &I) {
         OffsVal = DAG.getSExtOrTrunc(OffsVal, dl, N.getValueType());
 
-        N = DAG.getNode(ISD::ADD, dl, N.getValueType(), N, OffsVal, Flags);
+        N = DAG.getNode(AddOpcode, dl, N.getValueType(), N, OffsVal, Flags);
```
getMemBasePlusOffset?

rgwott replied:
Done.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -4293,6 +4293,12 @@ void SelectionDAGBuilder::visitGetElementPtr(const User &I) {
   GEPNoWrapFlags NW = cast<GEPOperator>(I).getNoWrapFlags();
+  unsigned int AddOpcode = ISD::PTRADD;
+
+  if (!DAG.getTarget().shouldPreservePtrArith(
```
Use getMemBasePlusOffset below instead?

rgwott replied:
Done.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -4473,8 +4478,15 @@ void SelectionDAGBuilder::visitAlloca(const AllocaInst &I) {
   // an address inside an alloca.
   SDNodeFlags Flags;
   Flags.setNoUnsignedWrap(true);
-  AllocSize = DAG.getNode(ISD::ADD, dl, AllocSize.getValueType(), AllocSize,
```
getMemBasePlusOffset?

rgwott replied:
Done.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -9071,8 +9083,13 @@ bool SelectionDAGBuilder::visitMemPCpyCall(const CallInst &I) {
+    AddOpcode = ISD::ADD;
+  }
+  SDValue DstPlusSize =
+      DAG.getNode(AddOpcode, sdl, Dst.getValueType(), Dst, Size);
```
getMemBasePlusOffset?

rgwott replied:
Done.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -11169,9 +11186,14 @@ TargetLowering::LowerCallTo(TargetLowering::CallLoweringInfo &CLI) const {
+              CLI.DAG.getMachineFunction().getFunction())) {
+        AddOpcode = ISD::ADD;
+      }
+      SDValue Add = CLI.DAG.getNode(
```
getMemBasePlusOffset?

rgwott replied:
Done.

arichardson reviewed (commented):
Thanks for upstreaming this - I think we should use getMemBasePlusOffset in more places though instead of all the added checks.

I also think the commit message should include Co-authored-by: for @davidchisnall and @jrtc27 (I don't think I substantively touched any of the imported lines so no need to include me. EDIT: just checked the blame, and my only contribution here is an infinite combine fix so I don't mind either way).

arichardson commented:
ping? Is this still being worked on?

rgwott commented:
> ping? Is this still being worked on?

Hi. This fell down the priority list for me, but I should be able to reply to the comments and resolve the outstanding matters in the next weeks.

Commit:
Merge branch main into branch cpa

rgwott committed:
[SelectionDAG] Refactor use of getMemBasePlusOffset() where applicable

Reduce complexity of PTRADD usage logic by harnessing
getMemBasePlusOffset().
Not all uses are applicable, as ADD's order of operands might be inverted
(PTRADD should always have pointer first, offset second).

rgwott committed:
[SelectionDAG] Inverted args for PTRADD on getMemBasePlusOffset()

By allowing getMemBasePlusOffset() to know whether its use for regular
ADDs does not take pointer first and offset second, the generation of
PTRADD on enabled targets can be done correctly, with the arguments
inverted.

This modification is to avoid changing the generation of some ADDs,
thus requiring the rewrite of several tests for several architectures.

rgwott committed:
Fix minor lexical problems

rgwott commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -109,7 +109,7 @@ def SDTOther  : SDTypeProfile<1, 0, [SDTCisVT<0, OtherVT>]>; // for 'vt'.
 def SDTUnaryOp  : SDTypeProfile<1, 1, []>;                   // for bitconvert.
 
-def SDTPtrAddOp : SDTypeProfile<1, 2, [     // ptradd
+def SDTPtrAddOp : SDTypeProfile<1, 2, [  // ptradd
```
That was a leftover modification that should not have been pushed. Thanks for noticing.

rgwott commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2578,6 +2581,98 @@ SDValue DAGCombiner::foldSubToAvg(SDNode *N, const SDLoc &DL) {
+
+    // (ptradd (ptradd x, y), z) -> (ptradd x, (add y, z)) if:
+    //   * x is a null pointer; or
+    //   * y is a constant and z has one use; or
```
I appreciate the work put into figuring out all those conditions, and some of them I am not even totally knowledgeable about. However we did find test cases that broke some. There was the possibility that new nodes were introduced, then `visit()` was called, then the reassociation would ultimately be cancelled after `visit()` modified the DAG. This left the DAG in an inconsistent state. My call was to remove the offending conditions as I found no better solution.

jrtc27 replied:
Could you please share these test cases?

rgwott replied:
Sure. Our fuzzer found several cases that are essentially the same, so I only have record of this:
`#include <stdint.h>
#include <stdio.h>

uint8_t a[2][1][2] = {1, 1, 1, 1};
uint16_t b = 0;

void main() {
  if (a[1][b][b + 1])
    printf("hello\n");
}`
The corresponding IR is in the test cases included in this PR (`multidim`).

rgwott replied:
Of course I may be wrong and it could be a problem with my adaptation.

rgwott committed:
Satisfy undef deprecator

rgwott commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2578,6 +2581,98 @@ SDValue DAGCombiner::foldSubToAvg(SDNode *N, const SDLoc &DL) {
+    bool YIsConstant = DAG.isConstantIntBuildVectorOrConstantInt(Y);
+    bool ZIsConstant = DAG.isConstantIntBuildVectorOrConstantInt(Z);
+
+    // (ptradd (ptradd x, y), z) -> (ptradd (ptradd x, z), y) if:
```
I have been scratching my head because of this. I see your point and agree, although this is probably a rare corner case. If I understand correctly, the only choice we have is to remove this fold completely, as we cannot know the sign of the variable during codegen.

jrtc27 replied:
There may well be ways to deduce it in some cases, I'm not familiar enough with that side of SelectionDAG to know, but yes, in the general case the fold is incorrect, so unless you have that information it needs to go

rgwott replied:
It is removed now. This changes opportunities for pattern matching of course, so we end up using simpler ADDPTs instead of MSUBPTs in some places. That should continue to be semantically correct however.

rgwott replied:
I really considered doing that, but I never know what the reviewers might think! =D
Please check if that is fine. I quoted your comment but did not cite your name, I figured you would not want it in the code.

jrtc27 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2612,6 +2615,98 @@ SDValue DAGCombiner::foldSubToAvg(SDNode *N, const SDLoc &DL) {
+  EVT IntVT = N1.getValueType();
+  SDLoc DL(N);
+
+  // fold (ptradd und*f, y) -> und*f
```
IMO these should stay as undef as the check is a false positive, which can be bypassed as needed

rgwott committed:
Remove fold with unhandeable corner case

rgwott committed:
Add comment explaining removed fold for future work

rgwott committed:
Minor lexical fix

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -8061,10 +8062,20 @@ SDValue SelectionDAG::getMemBasePlusOffset(SDValue Base, TypeSize Offset,
+          this->getMachineFunction().getFunction())) {
+    return getNode(ISD::ADD, DL, BasePtrVT, Ptr, Offset, Flags);
+  } else {
+    if (Inverted) {
```
I am not keen on this inverted flag - I will work on fixing all tests that needs this.

rgwott replied:
Ack, thanks.

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -8061,10 +8062,20 @@ SDValue SelectionDAG::getMemBasePlusOffset(SDValue Base, TypeSize Offset,
   assert(Offset.getValueType().isInteger());
   EVT BasePtrVT = Ptr.getValueType();
-  return getNode(ISD::ADD, DL, BasePtrVT, Ptr, Offset, Flags);
+  if (!this->getTarget().shouldPreservePtrArith(
```
For CHERI it might be good to pass the BasePtrVT to shouldPreservePtrArith() as well, but I guess that change can be made later. So no need to fix this in this PR.

arichardson reviewed (commented):
I think this looks mostly good now - I will work on submitting some changes shortly so that the "Inverted flag" is not needed anymore.

arichardson commented:
It should be possible to drop the inverted flag once https://github.com/llvm/llvm-project/pull/125279 lands.

One more nit: maybe the commit message should link to https://developer.arm.com/documentation/ddi0602/2024-12/Base-Instructions/ADDPT--Add-checked-pointer- instead of the top-level instruction reference?

arichardson commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -1072,7 +1072,8 @@ class SelectionDAG {
                                const SDNodeFlags Flags = SDNodeFlags());
   SDValue getMemBasePlusOffset(SDValue Base, SDValue Offset, const SDLoc &DL,
-                               const SDNodeFlags Flags = SDNodeFlags());
+                               const SDNodeFlags Flags = SDNodeFlags(),
+                               const bool Inverted = false);
```
```suggestion
                               const SDNodeFlags Flags = SDNodeFlags());
```

arichardson commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -8061,10 +8062,20 @@ SDValue SelectionDAG::getMemBasePlusOffset(SDValue Base, TypeSize Offset,
+          this->getMachineFunction().getFunction())) {
+    return getNode(ISD::ADD, DL, BasePtrVT, Ptr, Offset, Flags);
+  } else {
+    if (Inverted) {
+      return getNode(ISD::PTRADD, DL, BasePtrVT, Offset, Ptr, Flags);
+    } else {
+      return getNode(ISD::PTRADD, DL, BasePtrVT, Ptr, Offset, Flags);
+    }
```
```suggestion
    return getNode(ISD::PTRADD, DL, BasePtrVT, Ptr, Offset, Flags);
```

arichardson reviewed (approved):
Now that #125279 has landed this LGTM once rebased and the "inverted flag" has been dropped and the `ISD::ADD` for alignment has been restored.

davemgreen commented on llvm/test/CodeGen/AArch64/cpa-selectiondag.ll:
```diff
@@ -0,0 +1,451 @@
+  ret ptr %incdec.ptr.i.i.i
+}
+
+define i64 @subi64(i64 %ptr, i32 %index) {
```
Are all these tests useful?

rgwott replied:
In the previous PR, SelectionDAG did not differentiate between ptr and integer arithmetic (which was wrong). This PR fixes that, and these tests serve to show precisely that.

rgwott committed:
Merge branch main into branch cpa

rgwott committed:
Remove getMemBasePlusOffset inversion logic after #125279

rgwott committed:
Remove inadequate use of getMemBasePlusOffset()

rgwott committed:
Remove obsolete AddedCompletixy in FEAT_CPA tablegen patterns

rgwott committed:
Modify comment in PTRADD declaration

rgwott committed:
Gate FEAT_CPA CodeGen behind -mcpa-codegen flag

rgwott commented on llvm/test/CodeGen/AArch64/cpa-selectiondag.ll:
```diff
@@ -0,0 +1,451 @@
+; RUN: llc -mtriple=aarch64 -verify-machineinstrs --mattr=+cpa -O0 -global-isel=0 -fast-isel=0 %s -o - 2>&1 | FileCheck %s --check-prefixes=CHECK-CPA-O0
+; RUN: llc -mtriple=aarch64 -verify-machineinstrs --mattr=+cpa -O3 -global-isel=0 -fast-isel=0 %s -o - 2>&1 | FileCheck %s --check-prefixes=CHECK-CPA-O3
```
It is interesting to have because sometimes the instruction selected is different, allowing us to exercise both. This did however get a little bit less useful with the removal of a SelectionDAG fold opportunity during the course of this PR. But I am in favour of keeping it.

rgwott commented on llvm/test/CodeGen/AArch64/cpa-selectiondag.ll:
```diff
@@ -0,0 +1,451 @@
+@array2 = external dso_local global [10 x %struct.my_type2], align 8
+
+define void @addpt1(i64 %index, i64 %arg) {
+; CHECK-CPA-O0-LABEL:    addpt1:
```
I am not sure why, could you explain your rationale?

davemgreen replied:
It helps keep the tests regular and easier to maintain. (If you don't do it now, someone will likely regenerate them in the future if they change). You might be able to use --check-prefixes=CHECK-CPA,CHECK-CPA-O0 if we expect the O0 and O3 codegen to be the same in some tests.

rgwott replied:
Done.

rgwott commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -10382,6 +10382,26 @@ let Predicates = [HasCPA] in {
+  // Rules to use CPA instructions in pointer arithmetic patterns which are not
+  // folded into loads/stores. The AddedComplexity serves to help supersede
+  // other simpler (non-CPA) patterns and make sure CPA is used instead.
+  let AddedComplexity = 20 in {
```
That is a remainder from the previous PR, in which SelectionDAG did not differentiate between ptr and integer arithmetic (which was wrong), and this made instr selection "correct". I removed it now, well spotted.

rgwott commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -401,7 +401,7 @@ def tblockaddress: SDNode<"ISD::TargetBlockAddress",  SDTPtrLeaf, [],
 def add        : SDNode<"ISD::ADD"       , SDTIntBinOp   ,
                         [SDNPCommutative, SDNPAssociative]>;
-def ptradd     : SDNode<"ISD::ADD"       , SDTPtrAddOp, []>;
+def ptradd     : SDNode<"ISD::PTRADD"    , SDTPtrAddOp, []>;
```
This should not be a problem because PTRADD falls back to ADD in any target that does not opt in to shouldPreservedPtrArith().

rgwott replied:
I investigated a bit about this AMDGPU occurrence. When building the compiler, the tablegen will indeed compile to include a ptradd pattern to select a certain instruction. However: (1) there are two consecutive entries, one for add and one for ptradd, that do the same thing, so they are equivalent, and (2) in compile time, the PTRADD node will never exist for that backend so the tablegen entry will never be selected (the one for ADD will instead). Thankfully this is not a problem.

davemgreen replied:
It would be good to update the existing uses. If the AMD ptradd line is unnecessary, then removing it should not be a problem. It looks like it might be used for gisel at the moment though?

ritter-x2a replied:
The [AMDGPU ptradd line](https://github.com/llvm/llvm-project/blob/a481452cd88acc180f82dd5631257c8954ed7812/llvm/lib/Target/AMDGPU/VOP3Instructions.td#L751) has an effect for global ISel, it shouldn't be removed. This ptradd SDNode was introduced as an equivalent to global ISel's G_PTR_ADD (as declared in [SelectionDAGCompat.td](https://github.com/llvm/llvm-project/blob/a481452cd88acc180f82dd5631257c8954ed7812/llvm/include/llvm/Target/GlobalISel/SelectionDAGCompat.td#L63)), to specify SDAG patterns that are auto-translated to global ISel patterns.
As far as I'm aware, it doesn't matter for that if `ptradd` uses `ISD::ADD` or `ISD::PTRADD`, so changing it as the PR currently does is fine.

rgwott commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1462,6 +1462,10 @@ enum NodeType {
   PATCHPOINT,
 
+  // PTRADD represents pointer arithmetic semantics, for those targets which
+  // benefit from that information.
```
Done.

arichardson commented:
ping @rgwott. Any chance you will have time to update this PR and address the outstanding comments?

rgwott commented:
@arichardson Yes, of course. It's just that this is low priority at work right now. I intend to come back to it in some days.

ritter-x2a commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -2602,6 +2605,100 @@ SDValue DAGCombiner::foldSubToAvg(SDNode *N, const SDLoc &DL) {
+        (YIsConstant && N0OneUse) || (N0OneUse && ZOneUse && !ZIsConstant)) {
+      SDValue Add = DAG.getNode(ISD::ADD, DL, IntVT, {Y, Z});
+
+      // Calling visit() can replace the Add node with ISD::DELETED_NODE if
+      // there aren't any users, so keep a handle around whilst we visit it.
+      HandleSDNode ADDHandle(Add);
+
+      SDValue VisitedAdd = visit(Add.getNode());
+      if (VisitedAdd) {
+        // If visit() returns the same node, it means the SDNode was RAUW'd, and
+        // therefore we have to load the new value to perform the checks whether
+        // the reassociation fold is profitable.
+        if (VisitedAdd.getNode() == Add.getNode())
+          Add = ADDHandle.getValue();
+        else
+          Add = VisitedAdd;
+      }
```
Is there a reason why this needs to be done by directly calling `visit` here and cleaning up after it, instead of calling `AddToWorklist(Add.getNode())` so that the Combiner's worklist algorithm takes care of processing the new node next?
That seems to be the way it's done in the other combiner rules.

arichardson replied:
That's a good question - I believe that code might be my fault. It just seemed like the easiest way to avoid the assertion we were seeing before.

I *believe* adding the Add to the worklist should result in the same end result with less code complexity.

arichardson commented:
> Hi @rgwott, I'm working on changes in the AMDGPU backend that would benefit from the PTRADD SelectionDAG opcode that this PR adds. I understand that this PR is low priority for you right now, so I'd like to suggest extracting the relevant changes to enable PTRADD nodes (without the ARM/CPA-specific code) into a separate PR so that we can get them merged sooner. Would you be okay with that? If so, I can create a new PR with the extracted changes (setting you as the author for the commit, so that your work is credited).

If you would like to credit the original author for the PTRADD nodd that would be @davidchisnall (and probably some parts by @jrtc27, although that might just be the combines). I don't see any problems in landing that part separately

ritter-x2a commented:
> If you would like to credit the original author for the PTRADD nodd that would be @davidchisnall (and probably some parts by @jrtc27, although that might just be the combines). I don't see any problems in landing that part separately

Sounds like a good use for GitHub's [co-authors](https://github.blog/news-insights/product-news/commit-together-with-co-authors/) feature (as suggested by the [LLVM dev policy](https://llvm.org/docs/DeveloperPolicy.html#commit-messages)).
So, @davidchisnall as `--author`, with @jrtc27 and @rgwott as `Co-authored-by`, and also you, @arichardson, since you contributed the infinite combine fix?

rgwott commented:
@ritter-x2a @arichardson Sure, thank you for doing that. Then I will update this one accordingly and I can finish with the reviewers' suggestions shortly.

rgwott committed:
Merge remote-tracking branch 'origin/main' into cpa

rgwott committed:
Autogenerate FEAT_CPA tests with utils/update_llc_test_checks.py

rgwott committed:
Fix minor typo

rgwott committed:
Fix botched merge from upstream

rgwott committed:
Fix remaining botched merge from upstream

rgwott committed:
Adapt FEAT_CPA SDAG/GISel to shouldPreservePtrArith in targetLowering

rgwott committed:
Fix missing newline

rgwott committed:
Remove user-facing option -mcpa-codegen and use backend option instead

rgwott commented on clang/include/clang/Driver/Options.td:
```diff
@@ -5025,6 +5025,11 @@ def msve_vector_bits_EQ : Joined<["-"], "msve-vector-bits=">, Group<m_aarch64_Fe
   HelpText<"Specify the size in bits of an SVE vector register. Defaults to the"
            " vector length agnostic value of \"scalable\". (AArch64 only)">;
+
+def mcpa_codegen : Flag<["-"], "mcpa-codegen">,
```
Done.

rgwott committed:
Remove now-superfluous HasCPACodegen() function

davemgreen reviewed (commented):
Thanks for the updates, this is quite a bit simpler now. As far as I can tell this looks OK to me.

rgwott committed:
Remove -global-isel-abort=1 from GlobalISel tests

rgwott committed:
Insert printf() call in CPA tests to prevent case from optimizing away

Test case was based on this snippet:

 #include <stdint.h>
 #include <stdio.h>
 uint8_t a[2][1][2] = {1, 1, 1, 1};
 uint16_t b = 0;
 void main() {
     if (a[1][b][b + 1])
     printf("hello\n");
 }

In this patch we add the printf() call to avoid the test case from being
optimized away.

rgwott committed:
Make (ADDPT|SUBPT) shift operand an i64 as opposed to i32

To avoid the trunc_imm. Representing it as an i64 seems like a more natural
type for something based on pointers, even if the exact type does not
matter due to the range.

rgwott commented on llvm/test/CodeGen/AArch64/cpa-globalisel.ll:
```diff
@@ -0,0 +1,723 @@
+; CHECK-CPA-O0-NEXT:    ret
+;
+; CHECK-CPA-O3-LABEL: multidim:
+; CHECK-CPA-O3:       // %bb.0: // %entry
```
This test case was based on the following snippet:
https://github.com/llvm/llvm-project/pull/105669#discussion_r1927110106
Therefore I added the printf() and it should be good.
Done.

rgwott commented on llvm/test/CodeGen/AArch64/cpa-globalisel.ll:
```diff
@@ -0,0 +1,723 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64 -verify-machineinstrs --mattr=+cpa -aarch64-use-featcpa-codegen=true -O0 -global-isel=1 -global-isel-abort=1 %s -o - 2>&1 | FileCheck %s --check-prefixes=CHECK-CPA-O0
```
Done.

jthackray reviewed (approved):
LGTM

ostannard reviewed (approved):
LGTM
